### PR TITLE
fuse constant padding into conv kernels

### PR DESCRIPTION
**Pull request by mbrookhart**
cc @masahi @csullivan 

Currently supports conv1d/2d/3d. I can probably do the same thing for conv transpose and pooling, but this is getting a little long, what do you guys think?

### CODE REVIEW

File: src/relay/transforms/simplify_expr.cc [hunk 1]
```diff
@@ -82,6 +82,101 @@ class SimplifyReshape : public SimplifyPattern {
...
+    conv_ = (conv1d_ || conv2d_ || conv3d_)({pad_, w_});
+    pattern_ = conv_;
+  }
+  template <typename T>
+  Attrs MakeConvAttrs(const Attrs& attrs, const Array<PrimExpr> padding) const {
+    const T* old_attrs = attrs.as<T>();
+    ICHECK(old_attrs);
+    auto new_attrs = make_object<T>();
+    Array<PrimExpr> combined_padding;
+    ICHECK(padding.size() == old_attrs->padding.size());
```
Comment (csullivan) [inline src/relay/transforms/simplify_expr.cc:107]: Consider adding an error message, 
```suggestion
    ICHECK(padding.size() == old_attrs->padding.size()) << "Number of dimensions to pad and convolution padding attributes should have the same extent";
```

File: src/relay/transforms/simplify_expr.cc [hunk 2]
```diff
@@ -82,6 +82,101 @@ class SimplifyReshape : public SimplifyPattern {
...
+                const Map<DFPattern, Array<Expr>>& node_map) const override {
+    const CallNode* call_node = post.as<CallNode>();
+    ICHECK(call_node);
+    auto pad = node_map[pad_][0];
+    const CallNode* pad_node = pad.as<CallNode>();
+    ICHECK(pad_node);
+    const PadAttrs* param = pad_node->attrs.as<PadAttrs>();
+    ICHECK(param);
+    if (param->pad_mode == "constant" && param->pad_value == 0.0) {
+      for (size_t i = 0; i < 2; ++i) {
```
Comment (csullivan) [inline src/relay/transforms/simplify_expr.cc:133]: Should we avoid assuming a tuple length of 2?

```suggestion
      for (size_t i = 0; i < param->pad_width.size(); ++i) {
```

**Review by csullivan — COMMENTED** [review]
Thanks @mbrookhart! 
Support for other operators that handle would be a nice follow on, but this handles the majority of problematic cases of padding happening explicitly from ONNX's use of auto_pad. On deeplabv3-mobilenetv2 this reduces the number of explicit `nn.pad` calls from `54` -> `11`. Those remaining are part of a space to depth transformation.

File: src/relay/transforms/simplify_expr.cc [hunk 3]
```diff
@@ -82,6 +82,119 @@ class SimplifyReshape : public SimplifyPattern {
...
+  template <typename T>
+  Attrs GetAttrs(const PadAttrs* param, const T* attrs) const {
+    ICHECK(param);
+    ICHECK(attrs);
+    ICHECK(attrs->data_layout.size() == param->pad_width.size())
+        << "Data Layout and padding attributes should have the same extent";
+
+    std::string data_layout = attrs->data_layout;
+    std::set<char> image_dims({'H', 'W', 'D'});
+    Array<PrimExpr> padding;
```
Comment (csullivan) [inline src/relay/transforms/simplify_expr.cc:134]: Nice. Consider a comment here about bailing out if padding is not on one of the spatial dimensions

**Review by csullivan — APPROVED** [review]
Wonderful! Thanks @mbrookhart

**Comment by csullivan** [comment]
Note that given the layout inference of the implementation, users should consider running SimplifyExpr prior to layout transformations via ConvertLayout or otherwise. Else it may be the case that a `layout_transform` will block the fusion.

**Review by masahi — APPROVED** [review]

**Comment by masahi** [comment]
Thanks @mbrookhart @csullivan

**Pull request merged**